README.md
# simple_url_preview 
[![likes](https://badges.bar/simple_url_preview/likes)](https://pub.dev/packages/simple_url_preview/score)
[![pub points](https://badges.bar/simple_url_preview/pub%20points)](https://pub.dev/packages/simple_url_preview/score)
[![popularity](https://badges.bar/simple_url_preview/popularity)](https://pub.dev/packages/simple_url_preview/score)

Flutter package to show url preview

![In Action](https://github.com/Amitbhave/flutter-simple-url-preview/blob/master/preview_in_action.png)

## Getting Started

This shows url preview of a URL. (Now migrated to null-safety).

Currently only supports [Open Graph Protocol](https://www.ogp.me/)

Please use latest version of the package.

## How to use ?

Add simple_url_preview to pubspec.yaml, and hit command 'flutter pub get'
```yaml
dependencies:
  ...
  simple_url_preview: ^3.0.1
```

#### 1) **Simple use:**
```dart
SimpleUrlPreview(
  url: 'https://pub.dev/',
),
```

#### 2) **Override preview height, padding.(Default and minimum possible height is 130):**
```dart
SimpleUrlPreview(
  url: 'https://pub.dev/',
  previewHeight: 200,
  previewContainerPadding: EdgeInsets.all(10),
),
```

#### 3) **Override background color:**

Default bgColor = Theme.of(context).primaryColor

```dart
SimpleUrlPreview(
  url: 'https://pub.dev/',
  bgColor: Colors.red,
),
```

#### 4) **Override title, description, and site style:**

Default titleStyle = TextStyle(
  fontWeight: FontWeight.bold,
  fontSize: 16,
  color: Theme.of(context).accentColor
)

Default descriptionStyle = TextStyle(
  fontSize: 14,
  color: Theme.of(context).accentColor
)

Default siteNameStyle = TextStyle(
  fontSize: 14,
  color: Theme.of(context).accentColor
)

```dart
SimpleUrlPreview(
  url: 'https://pub.dev/',
  titleStyle: TextStyle(
    fontSize: 16,
    fontWeight: FontWeight.bold,
    color: Colors.red,
  ),
  descriptionStyle: TextStyle(
    fontSize: 14,
    color: Theme.of(context).primaryColor,
  ),
  siteNameStyle: TextStyle(
    fontSize: 14,
    color: Theme.of(context).primaryColor,
  ),
),
```

#### 5) **If you want closable preview (Click on x to close the preview.):**
```dart
SimpleUrlPreview(
  url: 'https://pub.dev/',
  isClosable: true,
),
```

#### 6) **Override image loader color and title and description lines:**

Default and maximum title lines = 2 and description lines = 3.

```dart
SimpleUrlPreview(
  url: 'https://pub.dev/',
  titleLines: 1,
  descriptionLines: 2,
  imageLoaderColor: Colors.white,
),
```

#### 7) **Override onTap callback of the URL preview:**

By Default, will open URL in default browser.

```dart
SimpleUrlPreview(
  url: 'https://pub.dev/',
  onTap: () => print('Hello Flutter URL Preview'),
),
```

### Contribution:

Would :heart: to see any contributions.

### Appreciate:
If you liked my work, show some :heart: by :star: repo.

Also you can appreciate by

<p>
 <table style="border-spacing: 5px 10px;">

 <tr>
  <td>
<a href="https://www.buymeacoffee.com/amitbhave10"><img src="https://cdn.buymeacoffee.com/buttons/default-orange.png" alt="Buy Me A Coffee" style="max-width:90%;" width="200" height="60"></a>
</td>

  <td style="margin: 10px">
<a href="https://paypal.me/amitbhave10"><img src="https://www.paypalobjects.com/webstatic/mktg/Logo/pp-logo-200px.png" alt="PayPal Logo"
style="max-width:90%;" width="200" height="60">
 </td>
 </tr>
 </table>
</p> 


CHANGELOG.md
## [3.0.1] - 15/06/2021.

* Update readme

## [3.0.0] - 15/06/2021.

* Migrated to null-safety!
* CacheStore dependency removed as that package doesn't support null safety.

## [2.0.0] - 31/01/2021.

* Add configuration options: testStyle, descriptionStyle and siteNameStyle.
* Remove textColor attribute as no longer needed.

## [1.1.1] - 24/01/2021.

* Fix description lines overflow

## [1.1.0] - 02/10/2020.

* Cache http response using CacheStore
* URL preview image set to cover

## [1.0.0] - 18/09/2020.

* Add configurable onTap callback for onTap of URL preview
* Make URL preview container padding configurable
* Make image perfect square
* Minimum height of this package is 130 now(after removing default padding)
* Refactor code

## [0.1.11] - 20/08/2020.

* Refactor code
* Add color for error widget for Image

## [0.1.10] - 28/07/2020.

* Fix Theme.of() use in initState()

## [0.1.9] - 28/07/2020.

* Add back initState method

## [0.1.8] - 28/07/2020.

* Add component mounted check

## [0.1.7] - 28/07/2020.

* Update public API documentation

## [0.1.6] - 28/07/2020.

* Remove image load animation, and show plain icon

## [0.1.5] - 28/07/2020.

* Fix formatting, dartdoc issues

## [0.1.4] - 28/07/2020.

* Use cached_network_image to avoid re-fetching image for same url
* Make titleLines and descriptionLines configurable
* Add 5% width to image
* Show loader when image is still loading. Override imageLoaderColor to change its color.

## [0.1.3] - 27/07/2020.

* Update open issues url(homepage)

## [0.1.2] - 27/07/2020.

* Update readme

## [0.1.1] - 27/07/2020.

* Update package description

## [0.1.0] - 27/07/2020.

* Initial release.
* Provide url preview after validating url
* On click of preview, launch the url using [url_launcher](https://pub.dev/packages/url_launcher).


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:simple_url_preview/simple_url_preview.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Simple Url Preview Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        primaryColor: Colors.blue,
        accentColor: Colors.white,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Simple Url Preview Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, this.title}) : super(key: key);

  final String? title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  String _url = '';

  _onUrlChanged(String updatedUrl) {
    setState(() {
      _url = updatedUrl;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title!),
      ),
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          SimpleUrlPreview(
            url: _url,
            bgColor: Theme.of(context).accentColor,
            isClosable: true,
            titleLines: 2,
            descriptionLines: 3,
            imageLoaderColor: Colors.white,
            previewHeight: 150,
            previewContainerPadding: EdgeInsets.all(10),
            onTap: () => print('Hello Flutter URL Preview'),
            titleStyle: TextStyle(
              fontSize: 16,
              fontWeight: FontWeight.bold,
              color: Colors.red,
            ),
            descriptionStyle: TextStyle(
              fontSize: 14,
              color: Theme.of(context).primaryColor,
            ),
            siteNameStyle: TextStyle(
              fontSize: 14,
              color: Theme.of(context).primaryColor,
            ),
          ),
          Padding(
            padding: EdgeInsets.all(20),
            child: TextField(
              onChanged: (newValue) => _onUrlChanged(newValue),
              decoration: InputDecoration(
                hintText: 'Enter the url',
              ),
            ),
          ),
        ],
      ),
    );
  }
}


lib/widgets/preview_image.dart
import 'package:cached_network_image/cached_network_image.dart';
import 'package:flutter/material.dart';

/// Shows image of URL
class PreviewImage extends StatelessWidget {
  final String? _image;
  final Color? _imageLoaderColor;

  PreviewImage(this._image, this._imageLoaderColor);

  @override
  Widget build(BuildContext context) {
    if (_image != null) {
      return Padding(
        padding: EdgeInsets.all(10),
        child: CachedNetworkImage(
          imageUrl: _image!,
          fit: BoxFit.cover,
          height: (MediaQuery.of(context).size.width -
                  MediaQuery.of(context).padding.top -
                  MediaQuery.of(context).padding.bottom) *
              0.25,
          errorWidget: (context, url, error) => Icon(
            Icons.error,
            color: _imageLoaderColor,
          ),
          progressIndicatorBuilder: (context, url, downloadProgress) => Icon(
            Icons.more_horiz,
            color: _imageLoaderColor,
          ),
        ),
      );
    } else {
      return SizedBox();
    }
  }
}


lib/simple_url_preview.dart
library simple_url_preview;

import 'package:collection/collection.dart';
import 'package:flutter/material.dart';
import 'package:html/dom.dart';
import 'package:html/parser.dart';
import 'package:http/http.dart';
import 'package:simple_url_preview/widgets/preview_description.dart';
import 'package:simple_url_preview/widgets/preview_image.dart';
import 'package:simple_url_preview/widgets/preview_site_name.dart';
import 'package:simple_url_preview/widgets/preview_title.dart';
import 'package:string_validator/string_validator.dart';
import 'package:url_launcher/url_launcher.dart';

/// Provides URL preview
class SimpleUrlPreview extends StatefulWidget {
  /// URL for which preview is to be shown
  final String url;

  /// Height of the preview
  final double previewHeight;

  /// Whether or not to show close button for the preview
  final bool? isClosable;

  /// Background color
  final Color? bgColor;

  /// Style of Title.
  final TextStyle? titleStyle;

  /// Number of lines for Title. (Max possible lines = 2)
  final int titleLines;

  /// Style of Description
  final TextStyle? descriptionStyle;

  /// Number of lines for Description. (Max possible lines = 3)
  final int descriptionLines;

  /// Style of site title
  final TextStyle? siteNameStyle;

  /// Color for loader icon shown, till image loads
  final Color? imageLoaderColor;

  /// Container padding
  final EdgeInsetsGeometry? previewContainerPadding;

  /// onTap URL preview, by default opens URL in default browser
  final VoidCallback? onTap;

  SimpleUrlPreview({
    required this.url,
    this.previewHeight = 130.0,
    this.isClosable,
    this.bgColor,
    this.titleStyle,
    this.titleLines = 2,
    this.descriptionStyle,
    this.descriptionLines = 3,
    this.siteNameStyle,
    this.imageLoaderColor,
    this.previewContainerPadding,
    this.onTap,
  })  : assert(previewHeight >= 130.0,
            'The preview height should be greater than or equal to 130'),
        assert(titleLines <= 2 && titleLines > 0,
            'The title lines should be less than or equal to 2 and not equal to 0'),
        assert(descriptionLines <= 3 && descriptionLines > 0,
            'The description lines should be less than or equal to 3 and not equal to 0');

  @override
  _SimpleUrlPreviewState createState() => _SimpleUrlPreviewState();
}

class _SimpleUrlPreviewState extends State<SimpleUrlPreview> {
  Map? _urlPreviewData;
  bool _isVisible = true;
  late bool _isClosable;
  double? _previewHeight;
  Color? _bgColor;
  TextStyle? _titleStyle;
  int? _titleLines;
  TextStyle? _descriptionStyle;
  int? _descriptionLines;
  TextStyle? _siteNameStyle;
  Color? _imageLoaderColor;
  EdgeInsetsGeometry? _previewContainerPadding;
  VoidCallback? _onTap;

  @override
  void initState() {
    super.initState();
    _getUrlData();
  }

  @override
  void didUpdateWidget(SimpleUrlPreview oldWidget) {
    super.didUpdateWidget(oldWidget);
    _getUrlData();
  }

  void _initialize() {
    _previewHeight = widget.previewHeight;
    _descriptionStyle = widget.descriptionStyle;
    _descriptionLines = widget.descriptionLines;
    _titleStyle = widget.titleStyle;
    _titleLines = widget.titleLines;
    _siteNameStyle = widget.siteNameStyle;
    _previewContainerPadding = widget.previewContainerPadding;
    _onTap = widget.onTap ?? _launchURL;
  }

  void _getUrlData() async {
    if (!isURL(widget.url)) {
      setState(() {
        _urlPreviewData = null;
      });
      return;
    }

    var response = await get(Uri.parse(widget.url));
    if (response.statusCode != 200) {
      if (!this.mounted) {
        return;
      }
      setState(() {
        _urlPreviewData = null;
      });
    }

    var document = parse(response.body);
    Map data = {};
    _extractOGData(document, data, 'og:title');
    _extractOGData(document, data, 'og:description');
    _extractOGData(document, data, 'og:site_name');
    _extractOGData(document, data, 'og:image');

    if (!this.mounted) {
      return;
    }

    if (data.isNotEmpty) {
      setState(() {
        _urlPreviewData = data;
        _isVisible = true;
      });
    }
  }

  void _extractOGData(Document document, Map data, String parameter) {
    var titleMetaTag = document
        .getElementsByTagName("meta")
        .firstWhereOrNull((meta) => meta.attributes['property'] == parameter);
    if (titleMetaTag != null) {
      data[parameter] = titleMetaTag.attributes['content'];
    }
  }

  void _launchURL() async {
    if (await canLaunch(Uri.encodeFull(widget.url))) {
      await launch(Uri.encodeFull(widget.url));
    } else {
      throw 'Could not launch ${widget.url}';
    }
  }

  @override
  Widget build(BuildContext context) {
    _isClosable = widget.isClosable ?? false;
    _bgColor = widget.bgColor ?? Theme.of(context).primaryColor;
    _imageLoaderColor =
        widget.imageLoaderColor ?? Theme.of(context).accentColor;
    _initialize();

    if (_urlPreviewData == null || !_isVisible) {
      return SizedBox();
    }

    return Container(
      padding: _previewContainerPadding,
      height: _previewHeight,
      child: Stack(
        children: [
          GestureDetector(
            onTap: _onTap,
            child: _buildPreviewCard(context),
          ),
          _buildClosablePreview(),
        ],
      ),
    );
  }

  Widget _buildClosablePreview() {
    return _isClosable
        ? Align(
            alignment: Alignment.topRight,
            child: IconButton(
              icon: Icon(
                Icons.clear,
              ),
              onPressed: () {
                setState(() {
                  _isVisible = false;
                });
              },
            ),
          )
        : SizedBox();
  }

  Card _buildPreviewCard(BuildContext context) {
    return Card(
      elevation: 5,
      color: _bgColor,
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
        children: [
          Container(
            width: (MediaQuery.of(context).size.width -
                    MediaQuery.of(context).padding.left -
                    MediaQuery.of(context).padding.right) *
                0.25,
            child: PreviewImage(
              _urlPreviewData!['og:image'],
              _imageLoaderColor,
            ),
          ),
          Expanded(
            child: Padding(
              padding: EdgeInsets.all(8),
              child: Column(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                crossAxisAlignment: CrossAxisAlignment.start,
                children: <Widget>[
                  PreviewTitle(
                      _urlPreviewData!['og:title'],
                      _titleStyle == null
                          ? TextStyle(
                              fontWeight: FontWeight.bold,
                              fontSize: 16,
                              color: Theme.of(context).accentColor,
                            )
                          : _titleStyle,
                      _titleLines),
                  PreviewDescription(
                    _urlPreviewData!['og:description'],
                    _descriptionStyle == null
                        ? TextStyle(
                            fontSize: 14,
                            color: Theme.of(context).accentColor,
                          )
                        : _descriptionStyle,
                    _descriptionLines,
                  ),
                  PreviewSiteName(
                    _urlPreviewData!['og:site_name'],
                    _siteNameStyle == null
                        ? TextStyle(
                            fontSize: 14,
                            color: Theme.of(context).accentColor,
                          )
                        : _siteNameStyle,
                  ),
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }
}
